### ISSUE TITLE
Allow to use environment variables in COPY --from

### ISSUE THREAD

**Issue by asfernandes** [moby/moby#34482]
Docker composes has the notion of project name (a prefix for the services). It may be set with env. var. $COMPOSE_PROJECT_NAME.

It would be good if we can use it with COPY --from inside Dockerfiles:

```
FROM ubuntu:xenial

COPY --from=${COMPOSE_PROJECT_NAME}_container /opt /opt
```

It gives error:

```
ERROR: Service '...' failed to build: invalid from flag value ${COMPOSE_PROJECT_NAME}_container: invalid reference format: repository name must be lowercase
```

**Comment by cpuguy83** [comment]
To what end, exactly?

This feature works like so:

```Dockerfile
FROM image1 AS name1
RUN do_stuff

FROM iamge2 AS name2
COPY --from=name1 /foo /bar
```

You can't copy from arbitrary containers/images.

**Comment by asfernandes** [comment]
As you name the project (prefix) name in docker-compose.yml, you currently need to replicate this name in Dockerfiles. So we have a name split into multiple places refering to the same thing and that is bad.

**Comment by cpuguy83** [comment]
Why do you have to replicate the name?
The dockerfile should be completely unaware of any naming and totally self-contained.

**Comment by asfernandes** [comment]
To use FROM to refer to another Dockerfile of the same set defined in docker-compose.

**Comment by cpuguy83** [comment]
That's not possible. The Dockerfile must be self-contained.

**Comment by asfernandes** [comment]
How can't it be possible if I'm doing it?

```
asfernandes@cheetah:/tmp/docker$ find .
.
./container1
./container1/Dockerfile
./container2
./container2/Dockerfile
./docker-compose.yml
asfernandes@cheetah:/tmp/docker$ 
asfernandes@cheetah:/tmp/docker$ 
asfernandes@cheetah:/tmp/docker$ cat ./container1/Dockerfile 
FROM ubuntu
asfernandes@cheetah:/tmp/docker$ 
asfernandes@cheetah:/tmp/docker$ cat ./container2/Dockerfile 
FROM ubuntu

COPY --from=temp_container1 /usr /tmp/usr
asfernandes@cheetah:/tmp/docker$ 
asfernandes@cheetah:/tmp/docker$ cat docker-compose.yml 
version: "2"

services:
    container1:
        build: ./container1

    container2:
        build: ./container2

asfernandes@cheetah:/tmp/docker$ 
asfernandes@cheetah:/tmp/docker$ COMPOSE_PROJECT_NAME=temp docker-compose build
Building container1
Step 1/1 : FROM ubuntu
 ---> 6a2f32de169d
Successfully built 6a2f32de169d
Successfully tagged temp_container1:latest
Building container2
Step 1/2 : FROM ubuntu
 ---> 6a2f32de169d
Step 2/2 : COPY --from=temp_container1 /usr /tmp/usr
 ---> Using cache
 ---> 43e6aaa55488
Successfully built 43e6aaa55488
Successfully tagged temp_container2:latest
```

I'm just saying it's nonsense to have to repeat the project name (temp), so I'm asking for a way to read it from env. var.

**Comment by asfernandes** [comment]
--from can use a image name, not only a thing defined inside the same Dockerfile.

**Comment by cpuguy83** [comment]
TIL

**Comment by tonistiigi** [comment]
@asfernandes You can work around it by defining a new stage.

```
FROM image-${witharg} AS stageforcopy

FROM alpine
COPY --from=stageforcopy foo bar
```

Expanding local arguments would be risky here as then we can't figure out the dependencies of the stages before pulling configurations for used images(something that https://github.com/moby/buildkit does for example for parallel builds) as they may change the `ARG` values.

We could expand to "FROM-args", the arguments that are in the beginning of a file and can be used in `FROM` commands but having this separation may be too confusing.

**Comment by thaJeztah** [comment]
> Expanding local arguments would be risky here as then we can't figure out the dependencies of the stages before pulling configurations for used images(something that https://github.com/moby/buildkit does for example for parallel builds) as they may change the ARG values.

@tonistiigi I was trying to envision the exact problem here. Given the following example;


```Dockerfile
# assume this is an image from Docker Hub
FROM busybox AS mybaseimage
ENV DOCKER_VERSION=18.09

FROM mybaseimage

# TAG is set from an environment variable from the current image ...
# .. which isn't known until that image (`mybaseimage`) is pulled from Docker Hub
ARG TAG=${DOCKER_VERSION}

# ...so the dependency for this `COPY` cannot be determined up-front
# however, the COPY cannot be executed until `FROM mybaseimage` is run,
# which would be needed anyway, so is that actually problematic?
# COPY --from=docker:${DOCKER_VERSION} /usr/local/bin/docker /usr/local/bin/
RUN echo COPY --from=docker:${DOCKER_VERSION} /usr/local/bin/docker /usr/local/bin/
```

(As commented inline); the `COPY` step cannot be executed until `FROM mybaseimage` has run,
so is it actually a problem that `TAG` is set dynamically from local arguments?


(I guess for https://github.com/moby/moby/issues/35018 at least it won't be a problem, because that won't influence the dependency graph)

**Comment by tonistiigi** [comment]
>  the COPY step cannot be executed until FROM mybaseimage has run,
so is it actually a problem that TAG is set dynamically from local arguments?

But we can already pull in the docker:18.09 in parallel while we are pulling busybox and executing the previous commands(and in more realistic case run all the commands from the dependant docker:version stage in parallel). That example can be easily converted to `FROM` args, but if you would remove the `ENV` command we don't know if `busybox` defines `DOCKER_VERSION` until we pull it.

**Comment by thaJeztah** [comment]
@vinayan3 build-args are scoped per build-stage; in your Dockerfile;

These args are declared _before_ the first FROM, thus in the "global" stage;


```Dockerfile
ARG API_BACKEND_CONTAINER="api:backend"
ARG API_STATIC_DIR="/var/www/static"
````

Which means they can be used in `FROM`;

```Dockerfile
FROM  $API_BACKEND_CONTAINER as source
```

A build-stage _starts_ at the `FROM` line;

```Dockerfile
FROM nginx:mainline-alpine
```

Which means that at this point (we're now inside the `FROM nginx:mainline-alpine` build stage), the `API_STATIC_DIR` is not defined:

```Dockerfile
ARG NGINX_ROOT=/usr/share/nginx/html/
COPY --from=source ${API_STATIC_DIR} ${NGINX_ROOT}
```

The `COPY` will thus look for an environment variable named `API_STATIC_DIR`, which is not set, so becomes an empty string.


You need to give to declare a build arg _inside_ that stage, to bring it into the stage's scope, and so that the stage has access to it;

Declaring the `ARG` without setting a value, will make it take the default value from the `ARG` defined in the global stage. The line below will default to `"/var/www/static"` (taken from the global `API_STATIC_DIR` build-arg), but can be overridden at build time using `--build-arg API_STATIC_DIR=<something>`;

```Dockerfile
ARG API_BACKEND_CONTAINER="api:backend"
ARG API_STATIC_DIR="/var/www/static"

FROM  $API_BACKEND_CONTAINER as source

FROM nginx:mainline-alpine
ARG API_STATIC_DIR
ARG NGINX_ROOT=/usr/share/nginx/html/
COPY --from=source ${API_STATIC_DIR} ${NGINX_ROOT}
```

**Comment by vinayan3** [comment]
@thaJeztah Thank you so much for the in-depth explanation. Is there plans to add a section for how ARGs interact with FROM stages into the documentation? It'd be probably very helpful to others. Only doc I found online was a little scant. I found a medium post which led me down the path of using variables in the FROM statement. https://medium.com/@tonistiigi/advanced-multi-stage-build-patterns-6f741b852fae

**Comment by thaJeztah** [comment]
@vinayan3 yes, build and the Dockerfile has a lot of flexibility, and not all of it may be fully documented. There's some information in the Dockerfile reference; https://docs.docker.com/engine/reference/builder/#scope

And a page on multistage builds https://docs.docker.com/develop/develop-images/multistage-build/

**Comment by dorfire** [comment]
> > the COPY step cannot be executed until FROM mybaseimage has run,
> > so is it actually a problem that TAG is set dynamically from local arguments?
> 
> But we can already pull in the docker:18.09 in parallel while we are pulling busybox and executing the previous commands

So the only benefit of NOT supporting args in `--from` is performance?

**Comment by tonistiigi** [comment]
Not only. There are two distinct sets of ARG definitions. Global ones used in `FROM` and local ones exposed to the `RUN` in stage scope, with a possibility to inherit a global one into the local stage. The reasons for that become obvious for any useful caching patterns in multi-stage builds.

If you consider ARG support in `--from` values, it makes much more sense that it would use the global args like `FROM`  does (the range of values `FROM` supports is also identical to `--from`).

Eg. a simple common example:

```
ARG ALPINE_VERSION=3.7
FROM alpine:$ALPINE_VERSION
...

FROM prod
COPY --form alpine:$ALPINE_VERSION foo bar
```

Same for the args in stage names, as `FROM` and `--from` takes same values they likely use same args. You don't want to expose `$ALPINE_VERSION` as an environment var inside the stage as it has no meaning in there.

We could do the above(exposing global args) with no technical issues. But it might be difficult to explain without explaining all the details of how/why `ARG` work like this. As shown above, we already support this functionality. It's a little more verbose to type but on the other hand, there is no way to misinterpret what it means, nor accidentally invalidate your cache or leak unnecessary variables into the image.

Basically, the only case that you can't do atm is where a hidden environment variable(not ARG) inside the `FROM` image is used as a variable replacement. And I'm not sure it is a good idea to allow it because it just hides what actually happens as part of your build from the reader of Dockerfile -  you can't predict what stages depend on each other before looking at the build output.

**Comment by bergkvist** [comment]
## In case you want to merge Docker images

I created a custom frontend (parser) for docker buildkit which adds an INCLUDE-keyword. This will copy in the entire file system, as well as import the environment variables. The PATH variable is merged in, while the other variables are just overwritten in the case of a conflict.

## Example Dockerfile
Notice the first line. This specifies that the Docker image I created (bergkvist/includeimage) should be used for parsing the Dockerfile
```Dockerfile
#syntax=bergkvist/includeimage
FROM alpine:3.12.0
INCLUDE rust:1.44-alpine3.12
INCLUDE python:3.8.3-alpine3.12
```

## Build with:
```
DOCKER_BUILDKIT=1 docker build -t myimage .
```

**Comment by greggwon** [comment]
I don't understand how it is not the case that each command line in a Dockerfile is first passed to variable expansion and then passed to syntax parsing.  This is how the unix shells have worked for decades and it is the most flexible and trivial implementation so that commands don't have to know anything about environment variable expansion in the own implementations.  Can this not be done?

**Comment by jules-ch** [comment]
envsubst is the best choice here no ?

export env variable and then build with your templated Dockerfile replacing env variable

```bash
envsubst < Dockerfile.template | docker build -f - .
```

**Comment by FibreFoX** [comment]
@jules-ch it goes beyond just environment variables, build arguments are not replaced too, so the only workaround seems to be the  named FROM workaround

(just stumbled upon this bug too)

**Comment by GCHQDeveloper314** [comment]
`podman build` does allow argument variables to be expanded. Although this appears to be entirely undocumented and potentially unintentional. I would suggest an error be raised when a user attempts this, to be clear it isn't supported.
